lib/ld-key-pair.js

```diff
@@ -0,0 +1,293 @@
+   * @param [options.passphrase]
+   *
+   * @returns {Promise<LDKeyPair>}
+   */
+  static async from(data, options) {
+    switch(data.keyType) {
+      case 'Ed25519VerificationKey2018':
+        return Ed25519KeyPair.from(data, options);
+
+      case 'RsaVerificationKey2018':
+      default:
+        return RSAKeyPair.from(data, options);
+    }
+  }
+
+  static async pbkdf2(password, salt, iterations, keySize) {
+    return new Promise((resolve, reject) => {
+      forge.pkcs5.pbkdf2(password, salt, iterations, keySize, (err, key) =>
+        err ? reject(err) : resolve(key));
+    });
+  }
+
+  /**
+   * @param owner {string} DID of key owner
+   */
+  publicNode({owner}) {
+    const publicNode = {
+      id: this.id,
+      type: this.keyType,
+      owner
+    };
+    this.addEncodedPublicKey(publicNode);
+    return publicNode;
+  }
+}
+
+class Ed25519KeyPair extends LDKeyPair {
+  static async generate(options) {
+    // currently node only
+    if(!options.injector.env.nodejs) {
+      throw new Error(
+        '"Ed25519VerificationKey2018" is not supported on this platform yet.');
+    }
+    const bs58 = require('bs58');
+    const chloride = require('chloride');
+    const keyPair = chloride.crypto_sign_keypair();
+
+    const keys = new Ed25519KeyPair({
+      publicKey: bs58.encode(keyPair.publicKey),
+      secretKey: bs58.encode(keyPair.secretKey),
+      ...options
+    });
+
+    return keys;
+  }
+
+  static async from(data, options) {
+    const keyPair = new Ed25519KeyPair({
+      publicKey: data.publicKeyBase58,
+      secretKey: data.secretKeyBase58,
+      id: data.id,
+      keyType: data.keyType,
+      ...options
+    });
+
+    return keyPair;
+  }
+
+  // publicKeyPem, publicKeyJwk, publicKeyHex, publicKeyBase64, publicKeyBase58
+  async export() {
+    let keyNode = {
+      id: this.id,
+      keyType: this.keyType,
+      publicKeyBase58: this.publicKey
+    };
+
+    return this.addEncryptedSecretKey(keyNode);
+  }
+
+  addEncodedPublicKey(publicKeyNode) {
+    publicKeyNode.publicKeyBase58 = this.publicKey;
+    return publicKeyNode;
+  }
+
+  async addEncryptedSecretKey(keyNode) {
+    if(this.passphrase !== null) {
+      keyNode.secretKeyJwe = await this.encrypt({secretKeyBase58: this.secretKey},
+        this.passphrase);
+    } else {
+      // no passphrase, do not encrypt private key
+      keyNode.secretKeyBase58 = this.secretKey;
+    }
+    return keyNode;
+  }
+
+  /**
+   * @param secretKey
+   * @param password
+   *
+   * @returns {Promise<JWE>}
+   */
+  async encrypt(secretKey, password) {
+    const keySize = 32;
+    const salt = forge.random.getBytesSync(32);
+    const iterations = 4096;
+    const key = await LDKeyPair.pbkdf2(password, salt, iterations, keySize);
+
+    const jweHeader = {
+      alg: 'PBES2-A128GCMKW',
+      enc: 'A128GCMKW',
+      jwk: {
+        kty: 'PBKDF2',
+        s: didUtil.encodeBase64Url(salt, {forge}),
+        c: iterations
+      }
+    };
```

Comment:
This JWE header looks like it needs work. I'll file a bug on it ... fixing it will be a breaking change, so we'll want to take care of it sooner rather than later (i.e. we'll want 0.7.x to be a short release that is quickly no longer supported).

Reply:
Yeah, agreed.